src/isa/instruction.h

```diff
@@ -0,0 +1,847 @@
+using IsField =
+    decltype(MaybeField::TokenIndex(), MaybeField::Ranges,
+             MaybeField::Apply(std::declval<const TokenizedSrcLine &>(),
+                               std::declval<Instr_T &>(),
+                               std::declval<FieldLinkRequest &>()),
+             MaybeField::Decode(std::declval<const Instr_T>(),
+                                std::declval<const Reg_T>(),
+                                std::declval<const ReverseSymbolMap &>(),
+                                std::declval<LineTokens &>()));
+
+template <typename MaybeBitRange>
+constexpr bool VerifyBitRange =
+    std::experimental::is_detected_v<IsBitRange, MaybeBitRange>;
+template <typename MaybeField>
+constexpr bool VerifyField =
+    std::experimental::is_detected_v<IsField, MaybeField>;
+
+template <template <typename...> class Op, typename...>
+struct VerifyValidTypes {};
+template <template <typename...> class Op, typename Type, typename... NextTypes>
+struct VerifyValidTypes<Op, Type, NextTypes...> {
+  static_assert(std::experimental::is_detected_v<Op, Type>, "Invalid type");
+
+  constexpr static VerifyValidTypes<Op, NextTypes...> VerifyRest{};
+};
+
+/** No-template, abstract class that describes a BitRange. */
+struct BitRangeBase {
+  constexpr BitRangeBase(unsigned _start, unsigned _stop, unsigned _N)
+      : start(_start), stop(_stop), n(_N) {}
+
+  const unsigned start, stop, n;
+
+  virtual unsigned width() const { return stop - start + 1; }
+  virtual Instr_T getMask() const { return vsrtl::generateBitmask(width()); }
+  virtual Instr_T apply(Instr_T value) const {
+    return (value & getMask()) << start;
+  }
+  virtual Instr_T decode(Instr_T instruction) const {
+    return (instruction >> start) & getMask();
+  }
+
+  virtual bool operator==(const BitRangeBase &other) const {
+    return n == other.n && start == other.start && stop == other.stop;
+  }
+  virtual bool operator<(const BitRangeBase &other) const {
+    if (n != other.n)
+      return (n < other.n);
+    return (start == other.start) ? stop < other.stop : start < other.start;
+  }
+};
+
+/** @brief A range of bits determined at compile-time
+ * NOTE: start/stop values for bitranges are inclusive..
+ * @param _start: Starting index of the range
+ * @param _stop: Ending index of the range (inclusive)
+ */
+template <unsigned _start, unsigned _stop, unsigned _N = 32>
+struct BitRange : public BitRangeBase {
+  static_assert(isPowerOf2(_N), "Bitrange N must be power of 2");
+  static_assert(_start <= _stop && _stop < _N, "invalid range");
+
+  constexpr BitRange() : BitRangeBase(_start, _stop, _N) {}
+
+  constexpr static unsigned N() { return _N; }
+  constexpr static unsigned Start() { return _start; }
+  constexpr static unsigned Stop() { return _stop; }
+  constexpr static unsigned Width() { return _stop - _start + 1; }
+  constexpr static Instr_T Mask() { return vsrtl::generateBitmask(Width()); }
+  constexpr static Instr_T Apply(Instr_T value) {
```

Comment:
After looking into it, I don't actually think C++20 would let us overload these functions. The problem is that they need to be static so that other classes (such as OpPart and Field) can call these functions without needing an instance of BitRange. Such as here, for example:

```c++
 struct Verify<FirstRange, SecondRange, OtherRanges...> {
    /// Assert that all BitRanges do not overlap with each other
    static_assert((FirstRange::ceStart() > SecondRange::ceStop() ||
                   FirstRange::ceStop() < SecondRange::ceStart()),
                  "BitRanges overlap with each other");
};
```

Let me know if you have any ideas to solve this, but I can't think of anything besides having duplicate functions for both static and instance contexts.

Reply:
Actually, I just had an idea. Do you think that something like this might be better? I'll use BitRange for this example.

The gist is that instead of duplicating each function so it can be called as static, there could be a single static function that returns an instance of the class (`getRange` in this example). Then any of the non-static functions can be called in a static context by using that function to get an instance.

```c++
struct BitRangeBase {
  constexpr BitRangeBase(unsigned _start, unsigned _stop, unsigned _N)
      : start(_start), stop(_stop), n(_N) {}

  const unsigned start, stop, n;

  unsigned width() const { return stop - start + 1; }
  Instr_T getMask() const {
    return generateMask(width());
  }
  Instr_T apply(Instr_T value) const {
    return (value & getMask()) << start;
  }
  Instr_T decode(Instr_T instruction) const {
    return (instruction >> start) & getMask();
  }

  bool operator==(const BitRangeBase &other) const {
    return n == other.n && start == other.start && stop == other.stop;
  }
  bool operator<(const BitRangeBase &other) const {
    if (n != other.n)
      return (n < other.n);
    return (start == other.start) ? stop < other.stop : start < other.start;
  }
};

template <unsigned _start, unsigned _stop, unsigned _N = 32>
struct BitRange : public BitRangeBase {
  static_assert(_start <= _stop && _stop < _N, "invalid range");

  constexpr BitRange() : BitRangeBase(_start, _stop, _N) {}

  constexpr static BitRangeBase getRange() {
    return BitRange();
  }
};
```

I think this would be better, as we wouldn't need to duplicate any functions. We wouldn't need to prefix any functions with `constexpr` either, since there won't be any duplicates. Please let me know what you think about this.

Reply:
Something like this? https://godbolt.org/z/Y5Y3jKKfc - wouldn't you still (as the error indicates) need to mark the various functions constexpr to actually use the function as `constexpr` even if you constructed the range in a constexpr manner?

Reply:
Apologies, I forgot to add constexpr to the BitRangeBase functions. Here's the fixed example: [https://godbolt.org/z/6oq88vzch](url)

In my previous comment, I just meant that we won't need to prefix the function names with `constexpr`/`ce` since there won't be any duplicate function names. The functions would still be defined as constexpr functions, they just won't need it in the function name.

Reply:
Awesome!